### PR TITLE
Added initial type definitions for usage in Typescript

### LINKED ISSUE #586: Add @types for typescript
Add @types for typescript

### PR DESCRIPTION
**Pull request by Shahaed**
Resolves elastic/apm-agent-nodejs#586 

Adding initial type declarations for typescript and updating documentation. Not everything is perfectly typed (there are a decent amount of `any`). More explicit typing suggestions are welcome!

### Checklist

- [x] Implement code
- [x] Update documentation

### CODE REVIEW

**Comment by watson** [comment]
Thanks for taking this one on 💯 

Two questions:

I don't know much about TypeScript, so this might be a stupid question, but `index.js` is exporting an instance of the `Agent` class. In your `index.d.ts` file, you export both each function by it self _and_ the `Agent` class.

Wouldn't it be possible to define the `Agent` class internally in `index.d.ts` (similar to how defined the `AgentOptions` interface) and then export an instance of that class instead?

I was also wondering if it would be feasible to use the specific types for the function arguments instead of the `any` type. That would probably help TypeScript users quite a lot as well 😃

**Comment by Shahaed** [comment]
@watson I simplified the `index.d.ts` file (following the example from [express](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/express/index.d.ts) ).

I typed the function best I could from the docs. If you don't mind, could you do a code review and point out which `any` types could be better (and tell me the types lol)? :smile:

File: index.d.ts [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        start(opts?: AgentOptions): Agent;
+        isStarted(): boolean;
+        addFilter(filter: (payload: any) => any): void;
+        setUserContext(context: Context): boolean;
+        setCustomContext(context: any): boolean;
+        setTag(name: string, value: string): any;
+        addTags(tags: any): any;
+        captureError(error: Error | string | object, options?: object, callback?: (err?: any) => any): void;
+        startTransaction(name?: string, type?: string): any;
```
Comment (Qard) [inline index.d.ts]: It would be good to implement and return a `Transaction` type here.

File: index.d.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        endTransaction(result?: any): any;
+        setTransactionName(name: string): any;
+        startSpan(name?: string, type?: any): any;
```
Comment (Qard) [inline index.d.ts]: It would be good to implement and return a `Span` type here. Also, `type` should be `string`.

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        isStarted(): boolean;
+        addFilter(filter: (payload: any) => any): void;
+        setUserContext(context: Context): boolean;
+        setCustomContext(context: any): boolean;
+        setTag(name: string, value: string): any;
+        addTags(tags: any): any;
+        captureError(error: Error | string | object, options?: object, callback?: (err?: any) => any): void;
+        startTransaction(name?: string, type?: string): any;
+        endTransaction(result?: any): any;
+        setTransactionName(name: string): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        setTransactionName(name: string): void;
```

File: index.d.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        start(opts?: AgentOptions): Agent;
+        isStarted(): boolean;
+        addFilter(filter: (payload: any) => any): void;
+        setUserContext(context: Context): boolean;
+        setCustomContext(context: any): boolean;
+        setTag(name: string, value: string): any;
+        addTags(tags: any): any;
+        captureError(error: Error | string | object, options?: object, callback?: (err?: any) => any): void;
+        startTransaction(name?: string, type?: string): any;
+        endTransaction(result?: any): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        endTransaction(result?: any): void;
```

File: index.d.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        currentTransaction: any;
+        logger: any;
+
+        start(opts?: AgentOptions): Agent;
+        isStarted(): boolean;
+        addFilter(filter: (payload: any) => any): void;
+        setUserContext(context: Context): boolean;
+        setCustomContext(context: any): boolean;
+        setTag(name: string, value: string): any;
+        addTags(tags: any): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        addTags(tags: any): boolean;
```

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        captureError(error: Error | string | object, options?: object, callback?: (err?: any) => any): void;
```
Comment (Qard) [inline index.d.ts]: The `err` argument of the callback should only ever be an `Error` instance, and the callback return value is not used, so you could make that stricter:

```
captureError(error: Error | string | object, options?: object, callback?: (err?: Error) => void): void;
```

Also, `options` is optional, so another valid signature is:

```
captureError(error: Error | string | object, callback?: (err?: Error) => void): void;
```

File: index.d.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+        startTransaction(name?: string, type?: string): any;
+        endTransaction(result?: any): any;
+        setTransactionName(name: string): any;
+        startSpan(name?: string, type?: any): any;
+        handleUncaughtExceptions(callback?: (err: Error) => any): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        handleUncaughtExceptions(callback?: (err: Error) => void): void;
```

File: index.d.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        flush(callback?: any): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        flush(callback?: (err?: Error) => void): void;
```

File: index.d.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+        currentTransaction: any;
+        logger: any;
+
+        start(opts?: AgentOptions): Agent;
+        isStarted(): boolean;
+        addFilter(filter: (payload: any) => any): void;
+        setUserContext(context: Context): boolean;
+        setCustomContext(context: any): boolean;
+        setTag(name: string, value: string): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        setTag(name: string, value: string): boolean;
```

File: index.d.ts [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+declare namespace a {
+
+    interface Agent {
+
+        currentTransaction: any;
+        logger: any;
+
+        start(opts?: AgentOptions): Agent;
+        isStarted(): boolean;
+        addFilter(filter: (payload: any) => any): void;
```
Comment (Qard) [inline index.d.ts]: Might be worthwhile for us to describe the payload types here too.

File: index.d.ts [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+        addTags(tags: any): any;
+        captureError(error: Error | string | object, options?: object, callback?: (err?: any) => any): void;
+        startTransaction(name?: string, type?: string): any;
+        endTransaction(result?: any): any;
+        setTransactionName(name: string): any;
+        startSpan(name?: string, type?: any): any;
+        handleUncaughtExceptions(callback?: (err: Error) => any): any;
+        flush(callback?: any): any;
+        lambda(handler: any): any;
+        lambda(type: any, handler: any): any;
```
Comment (Qard) [inline index.d.ts]:
```suggestion
        lambda(type: string, handler: any): any;
```

**Comment by watson** [comment]
@Shahaed would you still be interested in talking on this PR? If not, that's totally fine of course. If so, we'll probably take a look later when we have more bandwidth 😃

**Comment by Shahaed** [comment]
@watson Yeah, I can finish this PR! I've been AFK for a bit, sorry for the late response

**Comment by watson** [comment]
I just discovered that the definitions on DefinitelyTyped are very out of date (#917), so it would be good to land this ASAP. If you like, I can take over the PR :)

**Comment by watson** [comment]
I'm landing this in #926 instead. Hopefully that will land soon which should make this obsolete. So I'll close this for now. Thanks so much for you help to get this started 😊

**Pull request closed without merging**